### PR TITLE
Upgrade

### PR DESCRIPTION
**Pull request by khevse**

### CODE REVIEW

File: pkg/metric/provider.go [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+func (p *Provider) SuccessInc() {
+	p.success.Inc()
+}
+
+func (p *Provider) FailsInc() {
+	p.fails.Inc()
+}
+
+func (p *Provider) PushesInc(t time.Time) {
+	p.io.Observe(float64(time.Since(t).Nanoseconds()))
```
Comment (Badmoonz) [inline pkg/metric/provider.go]: в прометеус нужно писать в секундах

и лучше наверное использовать  встроенный таймер, так не ошиблись бы
timer := prometheus.NewTimer(p.io)
defer timer.ObserveDuration()

File: pkg/converter/api2ans/request.go [hunk 2]
```diff
@@ -0,0 +1,252 @@
...
+
+		topicList = append(topicList, list...)
+	}
+
+	// check config topic by certificate
+	if topic != "" {
+		var exist bool
+		for i := range topicList {
+			if strings.Compare(topicList[i], topic) == 0 {
+				exist = true
```
Comment (alexboyko) [inline pkg/converter/api2ans/request.go]: can break the cycle

File: pkg/provider/ans/client.go [hunk 3]
```diff
@@ -0,0 +1,218 @@
...
+	"github.com/pkg/errors"
+	"golang.org/x/net/http2"
+)
+
+type Client struct {
+	client         *http.Client
+	endpointPrefix string
+	certTLS        tls.Certificate
+	sandbox        bool
+	sendTries      int
```
Comment (alexboyko) [inline pkg/provider/ans/client.go]: retries is more suitable here

File: pkg/provider/ans/client.go [hunk 4]
```diff
@@ -0,0 +1,218 @@
...
+	existVoIP      bool
+}
+
+func New(certTLS *tls.Certificate, isSandbox bool, sendTries int, timeout time.Duration) (*Client, error) {
+
+	hasDevelopCert, err := ExistOID(certTLS, OidPushDevelop)
+	if err != nil {
+		return nil, errors.Wrap(err, "check develop certificate type")
```
Comment (alexboyko) [inline pkg/provider/ans/client.go]: errors should describe what went wrong like could not check certificate development oid

File: pkg/provider/ans/client.go [hunk 5]
```diff
@@ -0,0 +1,218 @@
...
+
+	sandbox := hasDevelopCert && (!hasProductionCert || isSandbox)
+
+	endpointPrefix := "https://api.push.apple.com"
+	if sandbox {
+		endpointPrefix = "https://api.development.push.apple.com"
+	}
+	endpointPrefix += "/3/device/"
+
+	if sendTries <= 0 {
```
Comment (alexboyko) [inline pkg/provider/ans/client.go]: if it's negative it's better to throw error, if it's zero, we can use default

File: pkg/provider/ans/client.go [hunk 6]
```diff
@@ -0,0 +1,218 @@
...
+
+func (c *Client) Certificate() tls.Certificate {
+	return c.certTLS
+}
+
+func (c *Client) Sandbox() bool {
+	return c.sandbox
+}
+
+func (c *Client) ExistVoIP() bool {
```
Comment (alexboyko) [inline pkg/provider/ans/client.go]: exists

File: pkg/provider/gcm/client.go [hunk 7]
```diff
@@ -0,0 +1,147 @@
...
+// Client (legacy/gcm)
+// https://firebase.google.com/docs/cloud-messaging/http-server-ref
+// Legacy FCM/GCM API (https://firebase.google.com/docs/cloud-messaging/migrate-v1):
+// 1. copy server key from: https://console.firebase.google.com/project/_/settings/cloudmessaging/
+// 2. add to request header: Authorization:key=<server key>
+type Client struct {
+	client *http.Client
+
+	// count send tries
+	sendTries int
```
Comment (alexboyko) [inline pkg/provider/gcm/client.go]: retries

File: pkg/provider/helper.go [hunk 8]
```diff
@@ -0,0 +1,48 @@
...
+	"errors"
+	"net/http"
+)
+
+var (
+	ErrInternalServerError = errors.New("remote push server: internal error")
+	ErrServiceUnavailable  = errors.New("remote push server: service unavailable")
+)
+
+func SendWithRetry(tries int, send func() (statusCode int, _ error)) error {
```
Comment (alexboyko) [inline pkg/provider/helper.go]: tries -> attempt

Comment (alexboyko) [inline pkg/provider/helper.go]: tries -> maxRetries

File: pkg/provider/helper_test.go [hunk 9]
```diff
@@ -0,0 +1,77 @@
...
+	"errors"
+	"net/http"
+	"testing"
+
+	"github.com/stretchr/testify/require"
+)
+
+func TestSendWithRetry(t *testing.T) {
+
+	for tries := 0; tries < 4; tries++ {
```
Comment (alexboyko) [inline pkg/provider/helper_test.go]: tries -> attempt

File: pkg/worker/fcm/worker.go [hunk 10]
```diff
@@ -0,0 +1,156 @@
...
+		w.sendNotification,
+	)
+	if err != nil {
+		return nil, err
+	}
+
+	return w, nil
+}
+
+func (w *Worker) ExistVoIP() bool {
```
Comment (alexboyko) [inline pkg/worker/fcm/worker.go]: supports voip?

File: pkg/converter/api2ans/request.go (hunk 2 above)
Comment (khevse) [reply pkg/converter/api2ans/request.go]: done

File: pkg/provider/ans/client.go (hunk 3 above)
Comment (khevse) [reply pkg/provider/ans/client.go]: done

File: pkg/provider/ans/client.go (hunk 4 above)
Comment (khevse) [reply pkg/provider/ans/client.go]: done

File: pkg/provider/ans/client.go (hunk 5 above)
Comment (khevse) [reply pkg/provider/ans/client.go]: done

File: pkg/provider/ans/client.go (hunk 6 above)
Comment (khevse) [reply pkg/provider/ans/client.go]: done

File: pkg/provider/gcm/client.go (hunk 7 above)
Comment (khevse) [reply pkg/provider/gcm/client.go]: done

File: pkg/provider/helper.go (hunk 8 above)
Comment (khevse) [reply pkg/provider/helper.go]: done

File: pkg/provider/helper_test.go (hunk 9 above)
Comment (khevse) [reply pkg/provider/helper_test.go]: done

File: pkg/worker/fcm/worker.go (hunk 10 above)
Comment (khevse) [reply pkg/worker/fcm/worker.go]: done

**Pull request merged**